core/embed/rust/src/ui/model_tr/component/title.rs

```diff
@@ -0,0 +1,102 @@
+use crate::{
+    micropython::buffer::StrBuffer,
+    time::Instant,
+    ui::{
+        component::{Component, Event, EventCtx, Marquee, Never},
+        display,
+        geometry::{Offset, Rect},
+        model_tr::theme,
+    },
+};
+
+pub struct Title {
+    area: Rect,
+    title: StrBuffer,
+    marquee: Marquee,
+    centered: bool,
+}
+
+impl Title {
+    pub fn new(title: StrBuffer) -> Self {
+        Self {
+            title,
+            marquee: Marquee::new(title, theme::FONT_HEADER, theme::FG, theme::BG),
+            area: Rect::zero(),
+            centered: false,
+        }
+    }
+
+    pub fn with_title_centered(mut self) -> Self {
+        self.centered = true;
+        self
+    }
+
+    /// Display title/header at the top left of the given area.
+    /// Returning the painted height of the whole header.
+    pub fn paint_header_left(title: StrBuffer, area: Rect) -> i16 {
+        let text_heigth = theme::FONT_HEADER.text_height();
+        let title_baseline = area.top_left() + Offset::y(text_heigth);
+        display::text_left(
+            title_baseline,
+            title.as_ref(),
+            theme::FONT_HEADER,
+            theme::FG,
+            theme::BG,
+        );
+        text_heigth
+    }
+
+    /// Display title/header centered at the top of the given area.
+    /// Returning the painted height of the whole header.
+    pub fn paint_header_centered(title: StrBuffer, area: Rect) -> i16 {
+        let text_heigth = theme::FONT_HEADER.text_height();
+        let title_baseline = area.top_center() + Offset::y(text_heigth);
+        display::text_center(
+            title_baseline,
+            title.as_ref(),
+            theme::FONT_HEADER,
+            theme::FG,
+            theme::BG,
+        );
+        text_heigth
+    }
+}
+
+impl Component for Title {
+    type Msg = Never;
+
+    fn place(&mut self, bounds: Rect) -> Rect {
+        self.area = bounds;
+        self.marquee.place(bounds);
+        bounds
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        if !self.marquee.is_animating() {
```

Comment:
Seem these marquee events will be triggered even in case the title fits entirely.

We could do the calculation we do every time in `paint`
(`self.needs_marque = width > self.area.width()`) already in `place` and not do the `event` at all in case it is `false`

Reply:
https://github.com/trezor/trezor-firmware/pull/2726/commits/8043226ae84d4b6cb1720dc52b52960660f34fa0

but had to start with `true` otherwise it misses the first event which comes before the paint and does not start the animation. not sure it if doesn't defeat the purpose though

not sure this is the right way to start the animation either

Reply:
That is unfortunate, let's leave it like this for now, if it works - we can go back to it in further reviews with more people

Reply:
On a second thought - even if "first event which comes before the paint", we should be setting `self.needs_marque` in `place()` - which should probably be called before `event`. In `paint()`, we do not have to recompute it again

Reply:
thats good idea, done here https://github.com/trezor/trezor-firmware/commit/314eb86056fae2fff1eb043b05fcd16c7c6bf17d